tinygrad/state.py

```diff
@@ -80,7 +80,7 @@ def _rebuild_tensor_v2(storage, storage_offset, size, stride, requires_grad, bac
 
     return ret.reshape(size)
 
-  intercept = {"HalfStorage": dtypes.float16, "FloatStorage": dtypes.float32, "IntStorage": dtypes.int32, "LongStorage": dtypes.int64, "_rebuild_tensor_v2": _rebuild_tensor_v2}
+  intercept = {"HalfStorage": dtypes.float16, "FloatStorage": dtypes.float32, "BFloat16Storage": dtypes.float16, "IntStorage": dtypes.int32, "LongStorage": dtypes.int64, "_rebuild_tensor_v2": _rebuild_tensor_v2}
```

Comment:
Err, this isn't right. dtypes.float16 isn't bfloat

Reply:
Right. I was trying to find a way how to hack bfloat16 support into tinygrad, but I failed.

Simply adding a new dtype to `helpers.dtype` did not do the trick, because the backend does not know this.

Please advise, how to proceed.

It seems the correct way would be to use float32 as storage and use code similar to this one to convert bfloat16 to float32:

https://github.com/ggerganov/llama.cpp/blob/294f424554c1599784ac9962462fc39ace92d8a5/convert.py#L338-L341

Reply:
Did a quick test locally and it matched up with pytorch but was pretty slow and double the memory footprint

```python
    if storage[1] == dtypes._bfloat16:
      temp = t[byte_offset:byte_offset+prod(size)].cast(dtypes.uint16)
      ret = Tensor((temp.numpy().astype(np.uint32) << 16).view(np.float32))
    else:
      ret = t[byte_offset:byte_offset+prod(size)].cast(storage[1])
```

Reply:
@dc-dc-dc I think that's better than nothing, we can always optimize later. Will you please push your changes into my PR branch? If that's not possible, will you please share your changes with me so I can update the PR?

Reply:
Not able to make a pr to your fork for some reason but I pushed the changes [here](https://github.com/dc-dc-dc/tinygrad/tree/llama-refactor)